client/rest/src/main/java/org/elasticsearch/client/Cancellable.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.client;
+
+import org.apache.http.client.methods.HttpRequestBase;
+
+import java.util.concurrent.CancellationException;
+
+/**
+ * Represents an operation that can be cancelled.
+ * Returned when executing async requests through {@link RestClient#performRequestAsync(Request, ResponseListener)}, so that the request
+ * can be cancelled if needed.
+ */
+public class Cancellable {
+
+    static final Cancellable NO_OP = new Cancellable(null) {
+        @Override
+        public void cancel() {
+        }
+
+        @Override
+        void runIfNotCancelled(Runnable runnable) {
+            throw new UnsupportedOperationException();
+        }
+    };
+
+    static Cancellable fromRequest(HttpRequestBase httpRequest) {
+        return new Cancellable(httpRequest);
+    }
+
+    private final HttpRequestBase httpRequest;
+
+    private Cancellable(HttpRequestBase httpRequest) {
+        this.httpRequest = httpRequest;
+    }
+
+    /**
+     * Cancels the on-going request that is associated with the current instance of {@link Cancellable}.
+     */
+    public synchronized void cancel() {
+        this.httpRequest.abort();
+    }
+
+    /**
+     * Executes some arbitrary code iff the on-going request has not been cancelled, otherwise it throws {@link CancellationException}.
+     * This is needed to guarantee that cancelling a request works correctly even in case {@link #cancel()} is called between different
+     * attempts of the same request. If the request has already been cancelled we don't go ahead, otherwise we run the provided
+     * {@link Runnable} which will reset the request and send the next attempt.
+     */
+    synchronized void runIfNotCancelled(Runnable runnable) {
```

Comment:
I don't think this method should be exposed in the `Cancellable` itself, as it is a custom business logic the `RestClient`
is only using (although these classes are all interleaved).
Also I don't think we need a runnable (see below),

What I would do is;

1.  expose the `isCancelled` in the interface 

```
public interface Cancellable {
    public void cancel()
    public boolean isCancelled()
}


```
make the InternalRequest implement Cancellable and return it
```
class InternalRequest implements Cancellable {
            public synchronized void cancel() {
               this.httpRequest.abort();
             }
         public synchronized void isCancelled() {
             this.httpRequest.isAborted();
        }
    }
```

If you don't like it then we could do a static factory method to create a Cancellable out of an InternalRequest.

Reply:
this method is not public, it's package private. Indeed it is internal, and that is why I did not expose it. It is just one way to handle concurrency to prevent cancel being a no-op when it's called between different attempts of the same request. I don't think we should address the complexity of retries by exposing isCancelled to users. Also, users can check isCancelled and then call cancel, but other operations can sneak in between, the point is to prevent that the request is aborted when we are done with one attempt but we haven't called execute on the next attempt (as effectively there is nothing to cancel because there is not future yet)